### PR TITLE
Arm64 arch

### PR DESCRIPTION
**Pull request by lipengfei28**
## Summary
1.goldfish:config pl031
2.arch/arm64: Modify the method for obtaining the CPU frequency
3.arch/arm64: Supports cluster PMU
4.qmeu_boot: add rpmsg syslog init for arm64 qemu chip
5.arch/arm64: GICv2 detection is compatible with different qemu versions
6.arch/arm64: Fixed GIC3 for goldfish platform
7.arch/arm64: Use serr to print fatal error messages.
8.arch/arm64: Fixed up_getusrsp getting stack Pointers

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



**No, this PR summary does not meet the NuttX requirements.**

**Here's why and how to improve it:**

**Summary:**

* **Insufficient Detail:** The summary provides a list of changes but lacks context and explanation. 
* **Missing Information:**  
    * **Why:**  The reason for each change (fix, update, feature) is not stated.  
    * **What:** A clear description of what part of the code is affected is needed. For example, "Modify the method for obtaining the CPU frequency" is vague. *Which* method? Where is it located?
    * **How:** How each change works is unclear.  For instance, how is the CPU frequency obtained now, and how will it be obtained after the change? 
* **No Issue References:**  Include any related NuttX or NuttX Apps issues or pull requests.

**Impact:**

* **Incomplete:** You need to address *all* impact categories, even if the answer is "NO."  Provide brief explanations for each "YES." 

**Testing:**

* **Insufficient Information:**
    * **Build Hosts:** List the specific operating systems, CPUs, and compiler versions used for testing.
    * **Targets:** Be specific about the architectures, boards, and configurations tested.
    * **Missing Logs:** You must include actual testing logs from *before* and *after* your changes.

**How to Write a Better Summary (Example):**

```
## Summary

This PR addresses [NuttX Issue #] by implementing support for the PL031 timer driver on the goldfish platform and enhancing CPU frequency acquisition and performance monitoring for the ARM64 architecture. 

Specifically, it:

1. **Adds pl031 support to goldfish:config:** This enables the use of the PL031 timer on the goldfish emulator.
2. **Modifies CPU frequency acquisition (arch/arm64):** Improves the accuracy and reliability of CPU frequency retrieval by [explain the new method and why it's better].
3. **Initializes perf for ARM64 QEMU (arch/arm64/qemu):** Enables performance monitoring capabilities within the QEMU environment for ARM64. 
4. **Adds cluster PMU support (arch/arm64):** Introduces support for cluster-level Performance Monitoring Units on ARM64, allowing for more granular performance analysis.

## Impact 

* **Is new feature added?** YES (pl031 support on goldfish, perf initialization on ARM64 QEMU, cluster PMU support) 
* **Impact on user:** YES. Users running on goldfish can now leverage the pl031 timer. Developers debugging on ARM64 QEMU can utilize perf tools.
* **Impact on build:** NO
* **Impact on hardware:** NO
* **Impact on documentation:** YES. Documentation will be updated to reflect the new pl031 driver configuration, CPU frequency acquisition method, perf usage, and cluster PMU support.
* **Impact on security:** NO
* **Impact on compatibility:** NO
* **Anything else to consider:** N/A

## Testing 

**Build Host(s):**
* Linux (Ubuntu 20.04), Intel Core i7, GCC 9.4.0

**Target(s):**
* ARM64 (QEMU virt machine), goldfish:config

**Testing logs before change:** 
[Insert logs showing the behavior before your changes]

**Testing logs after change:**
[Insert logs demonstrating the correct functionality after your changes] 
```

**Remember:** A well-written PR summary and impact assessment significantly helps reviewers understand and evaluate your changes.

**Comment by GUIDINGLI** [comment]
@pussuw 

Please review this PR

File: arch/arm64/src/common/Make.defs [hunk 1]
```diff
@@ -119,3 +119,7 @@ endif
 ifeq ($(CONFIG_ARM64_SEMIHOSTING_HOSTFS),y)
   CMN_CSRCS += arm64_hostfs.c
 endif
+
+ifeq ($(CONFIG_ARCH_HAVE_DEBUG),y)
```
Comment (pussuw) [inline arch/arm64/src/common/Make.defs:123]: You should make a config flag something like CONFIG_ARCH_USE_DEBUG (or similar) to give the user an option to leave the big module arm64_hwdebug.c out.

File: arch/arm64/src/common/arm64_hwdebug.c [hunk 2]
```diff
@@ -0,0 +1,1311 @@
...
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static uint64_t read_wb_reg(int reg, int n)
+{
+  uint64_t val = 0;
+
+  switch (reg + n)
+  {
```
Comment (pussuw) [inline arch/arm64/src/common/arm64_hwdebug.c:138]: Indent is wrong

File: arch/arm64/src/common/arm64_hwdebug.h [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+/* Debug register names. */
+
+#define AARCH64_DBG_REG_NAME_BVR        bvr
+#define AARCH64_DBG_REG_NAME_BCR        bcr
+#define AARCH64_DBG_REG_NAME_WVR        wvr
+#define AARCH64_DBG_REG_NAME_WCR        wcr
+
+/* Accessor macros for the debug registers. */
+
+#define AARCH64_DBG_READ(N, REG, VAL)            \
```
Comment (pussuw) [inline arch/arm64/src/common/arm64_hwdebug.h:232]: Formatting:

```
do 
  { 
  } 
while(0)
```

File: arch/arm64/src/common/arm64_pmuv3.h [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+#include <debug.h>
+#include <perf/arm_pmuv3.h>
+#include "arm64_arch.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define RETURN_READ_PMEVCNTRN(n) \
```
Comment (pussuw) [inline arch/arm64/src/common/arm64_pmuv3.h:36]: READ_PMEVCNTRN

File: drivers/perf/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+set(SRCS)
+
+if(CONFIG_SCHED_PERF_EVENTS)
+
+  if(CONFIG_ARCH_ARM64)
+    list(APPEND SRCS arm_pmu.c)
+    list(APPEND SRCS arm_pmuv3.c)
+  elseif(CONFIG_ARCH_ARM)
+    list(APPEND SRCS arm_pmu.c)
+    list(APPEND SRCS arm_pmuv3.c)
```
Comment (pussuw) [inline drivers/perf/CMakeLists.txt:29]: The same files for both platforms ?

File: include/perf/arm_pmu.h [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+typedef int (*armpmu_init_fn)(struct arm_pmu_s *);
+
+/****************************************************************************
+ * Inline Functions
+ ****************************************************************************/
+
+static inline bool cpumask_test_cpu(int cpu, int cpumask)
+{
+  return (bool)((1U << cpu) & cpumask);
```
Comment (pussuw) [inline include/perf/arm_pmu.h:147]: 1u

File: arch/arm64/src/common/arm64_fatal.c [hunk 7]
```diff
@@ -527,10 +527,10 @@ static int arm64_exception_handler(struct regs_context *regs)
 
   if (ret != 0)
     {
-      sinfo("CurrentEL: %s\n", el_str);
-      sinfo("ESR_ELn: 0x%" PRIx64 "\n", esr);
-      sinfo("FAR_ELn: 0x%" PRIx64 "\n", far);
-      sinfo("ELR_ELn: 0x%" PRIx64 "\n", elr);
+      serr("CurrentEL: %s\n", el_str);
```
Comment (pussuw) [inline arch/arm64/src/common/arm64_fatal.c:530]: You could even change the most critical messages to _alert(), like which exception occurred.

Also, ESR, FAR and ELR are absolutely vital information when taking a fatal exception where you absolutely CANNOT trust the stack contents (integer context).

File: arch/arm64/src/common/arm64_registerdump.c [hunk 8]
```diff
@@ -50,7 +50,7 @@
 uintptr_t up_getusrsp(void *regs)
 {
   struct regs_context *ptr = regs;
-  return ptr->regs[REG_X13];
+  return ptr->sp_elx;
```
Comment (pussuw) [inline arch/arm64/src/common/arm64_registerdump.c:53]: sp_elx is the kernel sp

**Comment by pussuw** [comment]
The PR is quite large, there are many unrelated changes in it. I would suggest splitting it into smaller chunks to make the review easier. With large PRs it is hard to follow all the changes.

At least the perf/pmu and debug / hwdebug changes should be 1 PR each.

File: arch/arm64/src/common/arm64_hwdebug.h (hunk 3 above)
Comment (lipengfei28) [reply arch/arm64/src/common/arm64_hwdebug.h:232]: done

File: arch/arm64/src/common/Make.defs (hunk 1 above)
Comment (lipengfei28) [reply arch/arm64/src/common/Make.defs:123]: to other pr

File: arch/arm64/src/common/arm64_hwdebug.c (hunk 2 above)
Comment (lipengfei28) [reply arch/arm64/src/common/arm64_hwdebug.c:138]: to other pr

File: arch/arm64/src/common/arm64_hwdebug.h (hunk 3 above)
Comment (lipengfei28) [reply arch/arm64/src/common/arm64_hwdebug.h:232]: to other pr

File: arch/arm64/src/common/arm64_pmuv3.h (hunk 4 above)
Comment (lipengfei28) [reply arch/arm64/src/common/arm64_pmuv3.h:36]: to other pr

File: drivers/perf/CMakeLists.txt (hunk 5 above)
Comment (lipengfei28) [reply drivers/perf/CMakeLists.txt:29]: to other pr

File: arch/arm64/src/common/arm64_fatal.c (hunk 7 above)
Comment (lipengfei28) [reply arch/arm64/src/common/arm64_fatal.c:530]: done

File: arch/arm64/src/common/arm64_registerdump.c (hunk 8 above)
Comment (lipengfei28) [reply arch/arm64/src/common/arm64_registerdump.c:53]: kernel build sp_el0 is user stack,
flat build sp_elx is user stack

File: include/perf/arm_pmu.h (hunk 6 above)
Comment (lipengfei28) [reply include/perf/arm_pmu.h:147]: to other pr

File: drivers/perf/CMakeLists.txt (hunk 5 above)
Comment (lipengfei28) [reply drivers/perf/CMakeLists.txt:29]: to other pr

File: arch/arm64/src/common/arm64_hwdebug.h (hunk 3 above)
Comment (lipengfei28) [reply arch/arm64/src/common/arm64_hwdebug.h:232]: to other pr

File: arch/arm64/src/common/arm64_hwdebug.c (hunk 2 above)
Comment (lipengfei28) [reply arch/arm64/src/common/arm64_hwdebug.c:138]: to other pr

File: arch/arm64/src/common/Make.defs (hunk 1 above)
Comment (lipengfei28) [reply arch/arm64/src/common/Make.defs:123]: to other pr

File: arch/arm64/src/common/arm64_pmuv3.h (hunk 4 above)
Comment (lipengfei28) [reply arch/arm64/src/common/arm64_pmuv3.h:36]: to other pr

**Review by pussuw — APPROVED** [review]
LGTM

**Pull request merged**